src/cplscheme/BaseCouplingScheme.cpp

```diff
@@ -725,62 +715,90 @@ void BaseCouplingScheme::advanceTXTWriters()
+  if(anyDataRequiresInitialization(receiveData)) {
+    _receivesInitializedData = true;
   }
 }
 
-void BaseCouplingScheme::timeWindowCompleted()
+bool BaseCouplingScheme::anyDataRequiresInitialization(BaseCouplingScheme::DataMap &dataMap) const
 {
-  PRECICE_TRACE(getTimeWindows(), getTime());
-  PRECICE_INFO("Time window completed");
-  setIsTimeWindowComplete(true);
-  setTimeWindows(getTimeWindows() + 1);
-  if (isCouplingOngoing()) {
-    PRECICE_DEBUG("Setting require create checkpoint");
-    requireAction(constants::actionWriteIterationCheckpoint());
+  /// @todo implement this function using https://en.cppreference.com/w/cpp/algorithm/all_any_none_of
+  for (DataMap::value_type &pair : dataMap) {
+    if (pair.second->requiresInitialization) {
+      return true;
+    }
   }
+  return false;
 }
 
-bool BaseCouplingScheme::maxIterationsReached()
+bool BaseCouplingScheme::receiveConvergence()
 {
-  if (not _isCoarseModelOptimizationActive) {
-    return _iterations == _maxIterations;
-  } else {
-    return _iterationsCoarseOptimization == _maxIterations;
+  PRECICE_ASSERT(doesFirstStep(), "For convergence information the receiving participant is always the first one.");
+  bool convergence;
+  _m2n->receive(convergence);
+  return convergence;
+}
+
+void BaseCouplingScheme::sendConvergence(bool convergence)
+{
+  PRECICE_ASSERT(not doesFirstStep(), "For convergence information the sending participant is never the first one.");
+  _m2n->send(convergence);
+}
+
+bool BaseCouplingScheme::accelerate()
+{
+  PRECICE_DEBUG("measure convergence of the coupling iteration");
+  bool convergence = measureConvergence();
+  // Stop, when maximal iteration count (given in config) is reached
+  if (maxIterationsReached())
+    convergence = true;
+
+  // coupling iteration converged for current time window. Advance in time.
+  if (convergence) {
+    if (getAcceleration()) {
+      _deletedColumnsPPFiltering = getAcceleration()->getDeletedColumns();
+      getAcceleration()->iterationsConverged(getAccelerationData());
+    }
+    newConvergenceMeasurements();
+    // no convergence achieved for the coupling iteration within the current time window
+  } else if (getAcceleration()) {
+    getAcceleration()->performAcceleration(getAccelerationData());
+  }
+
+  // extrapolate new input data for the solver evaluation in time.
+  if (convergence) {
+    extrapolateData(getAccelerationData()); // Also stores data
+  } else {                          // Store data for conv. measurement, acceleration, or extrapolation
+    for (DataMap::value_type &pair : getSendData()) {
+      if (pair.second->oldValues.size() > 0) {
+        pair.second->oldValues.col(0) = *pair.second->values;
+      }
+    }
+    for (DataMap::value_type &pair : getReceiveData()) {
+      if (pair.second->oldValues.size() > 0) {
+        pair.second->oldValues.col(0) = *pair.second->values;
+      }
+    }
```

Comment:
Looks buggy to me. For `SerialCouplingScheme` and `ParallelCouplingScheme` everything is fine, but for `MultiCouplingScheme` the functions `getSendData()` and `getReceiveData()` return nothing. 

I will try to handle this issue, when I introduce the `BiCouplingScheme`. For `MultiCouplingScheme` I will introduce an additional loop over `_sendDataVector`/`_receiveDataVector`:

```
    for (DataMap &sendData : _sendDataVector) {
      for (DataMap::value_type &pair : sendData) {
        if (pair.second->oldValues.size() > 0) {
          pair.second->oldValues.col(0) = *pair.second->values;
        }
      }
    }
    for (DataMap &receiveData : _receiveDataVector) {
      for (DataMap::value_type &pair : receiveData) {
        if (pair.second->oldValues.size() > 0) {
          pair.second->oldValues.col(0) = *pair.second->values;
        }
      }
    }
```

@uekerman does this look and sound reasonable?

Note: I put this functionality into a function called `storeData`. The name is not very good... any suggestions are welcome!